src/analytics/private/segments/user-segments-workflow.ts

```diff
@@ -0,0 +1,217 @@
+      outputPath: '$.Payload',
+    });
+    props.dataAPIRole.grantAssumeRole(segmentJobStatusFunc.grantPrincipal);
+    Bucket.fromBucketName(this, 'PipelineS3Bucket', props.pipelineS3Bucket).grantRead(segmentJobStatusFunc, `${props.segmentsS3Prefix}*`);
+
+    // Define Succeed and Fail end state
+    const succeedState = new Succeed(this, 'WorkflowEndState-Succeed');
+    const failState = new Fail(this, 'WorkflowEndState-Fail');
+
+    // Define choice state for workflow initialization
+    // If refresh schedule has expired, go to end state
+    const segmentJobInitChoice = new Choice(this, 'WorkflowChoice-SegmentJobInit')
+      .when(Condition.booleanEquals('$.scheduleIsExpired', true), succeedState)
+      .otherwise(stateMachineStatusTask);
+
+    // Connect segment job init task to the choice
+    segmentJobInitTask.next(segmentJobInitChoice);
+
+    // Define choice state for checking state machine status
+    // If state machine is idle, execute segment query. Otherwise, wait for 1 min and check again
+    const stateMachineStatusChoice = new Choice(this, 'WorkflowChoice-StateMachineStatus')
+      .when(Condition.stringEquals('$.stateMachineStatus', 'IDLE'), executeSegmentQueryTask)
+      .otherwise(new Wait(this, 'WorkflowWait-StateMachineStatus', {
+        time: WaitTime.duration(Duration.minutes(1)),
+      }).next(stateMachineStatusTask));
+
+    // Connect checking state machine status task to the choice
+    stateMachineStatusTask.next(stateMachineStatusChoice);
+
+    // Connect segment query execution task to checking job status task
+    executeSegmentQueryTask.next(segmentJobStatusTask);
+
+    // Define choice state for checking job status
+    const jobStatusChoice = new Choice(this, 'WorkflowChoice-SegmentJobStatus')
+      .when(Condition.stringEquals('$.jobStatus', SegmentJobStatus.COMPLETED), succeedState)
+      .when(Condition.stringEquals('$.jobStatus', SegmentJobStatus.FAILED), failState)
+      .otherwise(new Wait(this, 'WorkflowWait-CheckJobStatus', {
+        time: WaitTime.duration(Duration.seconds(15)),
+      }).next(segmentJobStatusTask));
+
+    // Connect checking job status task to the choice
+    segmentJobStatusTask.next(jobStatusChoice);
+
+    return new StateMachine(this, 'UserSegmentsStateMachine', {
+      definition: segmentJobInitTask,
+      logs: {
+        destination: createLogGroup(this, {
+          prefix: '/aws/vendedlogs/states/Clickstream/UserSegmentsStateMachine',
+        }),
+        level: LogLevel.ALL,
+      },
+      tracingEnabled: true,
+      comment: 'This state machine is responsible for creating and refreshing user segments.',
+    });
+  }
+
+  private constructNodejsFunction(name: string, policyStatements: PolicyStatement[], env: any = {}): IFunction {
+    return new SolutionNodejsFunction(this, `WorkflowLambda-${name}`, {
+      entry: join(lambdaRootPath, `${name}.ts`),
+      handler: 'handler',
+      memorySize: 1024,
+      timeout: Duration.minutes(15),
```

Comment:
Do we really need this long timeout value?

Reply:
No, the lambda will execute some DynamoDB, Redshift and S3 commands, which are unlikely to take more than 1 minute. Change the timeout to 1 minute.